# Permutations and Combinations
***
### Description
This project performs probability analysis using combinatorics methods to solve a modified experiment reminiscent of "Lady Tasting Tea". The problem uses 12 cups of tea: 6 of them with milk added first and 6 with tea added initially. The challenge is to calculate the probability that a person will correctly identify all 6 cups of milk added initially, as well as the probability of making one or two mistakes.

In [13]:
# Import lib
from math import comb

The comb function from the math module is used to compute binomial coefficients.

In [14]:
# Total number of combinations
total_combinations = comb(12, 6)
print(f"Total number of combinations: {total_combinations}")

Total number of combinations: 924


The total number of possible combinations to select 6 cups out of 12 is equale 924.
The total number of combinations shows how many possible sets can be made from 12 cups by selecting 6. This will allow to calculate probabilities for specific situations.

In [15]:
# The probability of guessing all 6 cups
favorable_combinations_all_correct = 1
probability_all_correct = favorable_combinations_all_correct / total_combinations
print(f"The probability of guessing all 6 cups: {probability_all_correct:.10f}")

The probability of guessing all 6 cups: 0.0010822511


The probability that all 6 cups will be guessed is calculated. Since there is only one combination where all the cups are correctly guessed, the probability is defined as the ratio of one favorable case to the total number of combinations.

The probability of guessing all 6 cups is very low, which is confirmed by the result. This shows that randomly guessing all the correct cups is almost impossible.

In [16]:
# Number of Ways to Guess Exactly 5 Correct Cups
ways_five_correct = comb(6, 5) * comb(6, 1)
print(f"Number of Ways to Guess Exactly 5 Correct Cups: {ways_five_correct}")

Number of Ways to Guess Exactly 5 Correct Cups: 36


It calculates the number of ways in which you can guess exactly 5 correct cups out of 6. First, choose 5 correct cups out of 6 and then add one wrong cup out of the remaining 6.

The number of ways to guess the 5 correct cups is much greater than the number of ways to guess all 6 correct ones. This makes it easier to calculate the probability for situations like this.

In [17]:
# Probability of making no more than one mistake
favorable_combinations_at_most_one_error = favorable_combinations_all_correct + ways_five_correct
probability_at_most_one_error = favorable_combinations_at_most_one_error / total_combinations
print(f"Probability of making no more than one mistake: {probability_at_most_one_error:.10f}")

Probability of making no more than one mistake: 0.0400432900


This block of code calculates the probability of making no more than one mistake - the sum of favorable cases for all correct guesses and for cases with one error, then divided by the total number of possible combinations.

The probability 0.0400432900 that no more than one mistake will be made is still low, making it unlikely in random attempts.

In [18]:
# The number of ways to guess exactly 4 correct cups
ways_four_correct = comb(6, 4) * comb(6, 2)
print(f"The number of ways to guess exactly 4 correct cups: {ways_four_correct}")

The number of ways to guess exactly 4 correct cups: 225


This block calculates the number of ways you can guess exactly 4 correct cups. Choose 4 correct cups out of 6 and two wrong ones from the rest.

There are many more ways to guess 4 correct cups than 5 or 6, resulting in an increase in favorable cases.

In [19]:
# The probability of making no more than two mistakes
favorable_combinations_at_most_two_errors = favorable_combinations_at_most_one_error + ways_four_correct
probability_at_most_two_errors = favorable_combinations_at_most_two_errors / total_combinations
print(f"The probability of making no more than two mistakes: {probability_at_most_two_errors:.10f}")

The probability of making no more than two mistakes: 0.2835497835


This block of code adds favorable cases for situations with no more than two errors and calculates the probability of such occurrences. The probability of making no more than two mistakes is slightly greater, but still remains low enough that it is unlikely with a random guess.

## Conclusion

In [20]:
# Conclusion on Making Two Mistakes
if probability_at_most_two_errors < 0.05:
    print("Two mistakes can be accepted because the probability of random success is still low.")
else:
    print("You shouldn't accept two mistakes, as the probability of random success is too high.")

You shouldn't accept two mistakes, as the probability of random success is too high.


The last block of code compares the probability of making no more than two mistakes with the 0.05 threshold and outputs a recommendation.

**Conclusion:**
If the probability is less than 0.05, two errors can be considered acceptable because the probability of random success remains low.

## Final Word
This project demonstrates how combinatorics and probability theory can be applied to the analysis of statistical experiments. He emphasizes the importance of understanding probabilities when making decisions based on statistical data.


# End
***